### PR TITLE
[in_app_purchase]make payment unified APIs

### PR DESCRIPTION
**Pull request by cyanglaz**
## Description

Add unified APIs for making a payment.
Accompanying making a payment, also add finishPurchase and consumePurchase APIs.
Add a `purchaseUpdatedStream` in the `InAppPurchaseConnection` for user to listen for purchase updating events.
Update example app for make purchase functionalities.

## Related Issues

flutter/flutter#26327
flutter/flutter#26328
flutter/flutter#26329
flutter/flutter#26331

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/in_app_purchase/lib/src/in_app_purchase_connection/in_app_purchase_connection.dart [hunk 1]
```diff
@@ -7,15 +7,243 @@ import 'dart:io';
...
+  /// The `applicationUserName`
+  /// The 'sandboxTesting' is only necessary to set to `true` for testing on iOS. The default value is `false`.
+  /// You can find more details on testing payments on iOS [here](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/StoreKitGuide/Chapters/ShowUI.html#//apple_ref/doc/uid/TP40008267-CH3-SW11).
+  /// You can find more details on testing payments on Android [here](https://developer.android.com/google/play/billing/billing_testing).
+  void makePayment(
+      {@required String productID,
+      String applicationUserName,
+      bool sandboxTesting = false});
+
+  /// Completes a purchase either after delivering the content or the purchase is failed. (iOS only).
```
Comment (mklim) [inline packages/in_app_purchase/lib/src/in_app_purchase_connection/in_app_purchase_connection.dart]: In general I'd like this API to be as consistent across platforms as possible. This is a non op on Play but I'd prefer if we asked users to call this in all cases, instead of checking for platform differences. This applies to a few of the API comments below.

File: packages/in_app_purchase/lib/src/in_app_purchase_connection/in_app_purchase_connection.dart [hunk 2]
```diff
@@ -189,60 +225,95 @@ abstract class InAppPurchaseConnection {
...
+      {@required PurchaseParam purchaseParam, bool autoConsume = true});
 
   /// Completes a purchase either after delivering the content or the purchase is failed. (iOS only).
   ///
-  /// Developer is responsible to complete every [PurchaseDetails] whose [PurchaseDetails.status] is [PurchaseStatus.purchased] or [[PurchaseStatus.error].
+  /// You are responsible to complete every [PurchaseDetails] whose [PurchaseDetails.status] is [PurchaseStatus.purchased] or [[PurchaseStatus.error].
   /// Completing a [PurchaseStatus.pending] purchase will cause exception.
   ///
-  /// This is a non-op on Android.
+  /// It throws an [Exception] on Android.
```
Comment (mklim) [inline packages/in_app_purchase/lib/src/in_app_purchase_connection/in_app_purchase_connection.dart]: Can you throw a more specific Exception type than this?

File: packages/in_app_purchase/lib/src/in_app_purchase_connection/product_details.dart [hunk 3]
```diff
@@ -25,6 +31,16 @@ class ProductDetails {
...
   final String price;
+
+  /// Points back to the `StoreKits`'s [SKProductWrapper] object that generated this [ProductDetails] object.
+  ///
+  /// This is null on Android.
+  final SKProductWrapper skProduct;
+
+  /// Points back to the `BillingClient1`'s [SkuDetailsWrapper] object that generated this [ProductDetails] object.
+  ///
+  /// This is null on Android.
```
Comment (mklim) [inline packages/in_app_purchase/lib/src/in_app_purchase_connection/product_details.dart]: nit: `iOS`

File: packages/in_app_purchase/example/lib/main.dart [hunk 4]
```diff
@@ -118,11 +149,64 @@ class _MyAppState extends State<MyApp> {
...
     }
 
+    // This loading previous purchases code is just a demo. Please do not use this as it is.
+    // In your app you should always verify the purchase data using the `verificationData` inside the [PurchaseDetails] object before trusting it.
+    // We recommend that you use your own server to verity the purchase data.
+    Map<String, PurchaseDetails> purchases = Map.fromEntries(
+        ((await connection.queryPastPurchases()).pastPurchases)
+            .map((PurchaseDetails purchase) {
+      print('restored: ${purchase.purchaseID}');
+      if (purchase.status == PurchaseStatus.pending) {
```
[the remark this answers is not in the document]
Comment (cyanglaz) [reply packages/in_app_purchase/example/lib/main.dart]: discussed offline, going to add this in the API with assertion.

File: packages/in_app_purchase/example/lib/main.dart [hunk 5]
```diff
@@ -40,7 +43,13 @@ class _MyAppState extends State<MyApp> {
           } else if (purchaseDetails.status == PurchaseStatus.purchased) {
             deliverProduct(purchaseDetails);
           }
-          InAppPurchaseConnection.instance.completePurchase(purchaseDetails);
+          if (Platform.isIOS) {
+            InAppPurchaseConnection.instance.completePurchase(purchaseDetails);
+          } else if (Platform.isAndroid) {
+            if (!Auto_Consume && purchaseDetails.productID == 'consumable') {
```
[the remark this answers is not in the document]
Comment (cyanglaz) [reply packages/in_app_purchase/example/lib/main.dart]: discussed offline, will add comment to this constant.

File: packages/cloud_firestore/example/android/settings.gradle [hunk 6]
```diff
@@ -12,4 +12,4 @@ plugins.each { name, path ->
     def pluginDirectory = flutterProjectRoot.resolve(path).resolve('android').toFile()
```
Comment (mklim) [inline packages/cloud_firestore/example/android/settings.gradle]: nit: Please revert the unrelated whitespace changes out of this PR.

File: packages/share/android/build.gradle [hunk 7]
```diff
@@ -35,7 +35,7 @@ rootProject.allprojects {
 apply plugin: 'com.android.library'
 
 android {
-    compileSdkVersion 28
+    compileSdkVersion 27
```
Comment (mklim) [inline packages/share/android/build.gradle]: nit: please revert this one.

File: packages/in_app_purchase/example/lib/main.dart [hunk 8]
```diff
@@ -120,26 +165,51 @@ class _MyAppState extends State<MyApp> {
...
-    Map<String, PurchaseDetails> purchases = Map.fromEntries(((await connection
-                .queryPastPurchases())
-            .pastPurchases)
-        .map((PurchaseDetails purchase) =>
-            MapEntry<String, PurchaseDetails>(purchase.productId, purchase)));
-
+    Map<String, PurchaseDetails> purchases = Map.fromEntries(
+        ((await connection.queryPastPurchases()).pastPurchases)
+            .map((PurchaseDetails purchase) {
+      print('restored productID: ${purchase.productID}');
```
Comment (mklim) [inline packages/in_app_purchase/example/lib/main.dart]: nit: I do think we'd be better off without all these logs, especially the ones just saying that a particular method has fired.
